zkevm-circuits/src/evm_circuit/execution/error_code_store.rs

```diff
@@ -0,0 +1,311 @@
+            LtGadget::construct(cb, MAXCODESIZE.expr(), memory_address.length());
+
+        // check must be one of CodeStoreOutOfGas or MaxCodeSizeExceeded
+        cb.require_in_set(
+            "CodeStoreOutOfGas or MaxCodeSizeExceeded",
+            code_store_gas_insufficient.expr() + max_code_size_exceed.expr(),
+            vec![1.expr(), 2.expr()],
+        );
+
+        let common_error_gadget = CommonErrorGadget::construct_with_return_data(
+            cb,
+            opcode.expr(),
+            4.expr(),
+            memory_address.offset(),
+            memory_address.length(),
+        );
+
+        Self {
+            opcode,
+            memory_address,
+            code_store_gas_insufficient,
+            max_code_size_exceed,
+            common_error_gadget,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _tx: &Transaction,
+        call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        let opcode = step.opcode().unwrap();
+        self.opcode
+            .assign(region, offset, Value::known(F::from(opcode.as_u64())))?;
+        // block.get_rws(step, index).stack_value());
+        let [memory_offset, length] = [0, 1].map(|i| block.get_rws(step, i).stack_value());
+        self.memory_address
+            .assign(region, offset, memory_offset, length)?;
+
+        self.code_store_gas_insufficient.assign(
+            region,
+            offset,
+            F::from(step.gas_left),
+            F::from(GasCost::CODE_DEPOSIT_BYTE_COST * length.as_u64()),
+        )?;
+
+        self.max_code_size_exceed.assign(
+            region,
+            offset,
+            F::from(MAXCODESIZE),
+            F::from(length.as_u64()),
+        )?;
+
+        self.common_error_gadget
+            .assign(region, offset, block, call, step, 4)?;
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use bus_mapping::circuit_input_builder::FixedCParams;
+    use eth_types::{
+        address, bytecode, evm_types::OpcodeId, geth_types::Account, Address, Bytecode, Word,
+    };
+
+    use lazy_static::lazy_static;
+    use mock::{eth, TestContext, MOCK_ACCOUNTS};
+
+    use crate::test_util::CircuitTestBuilder;
+
+    lazy_static! {
+        static ref CALLER_ADDRESS: Address = address!("0x00bbccddee000000000000000000000000002400");
+    }
+
+    const MAXCODESIZE: u64 = 0x6000u64;
+
+    fn run_test_circuits(ctx: TestContext<2, 1>) {
+        CircuitTestBuilder::new_from_test_ctx(ctx)
+            .params(FixedCParams {
+                max_rws: 4500,
+                ..FixedCParams::default()
+            })
+            .run();
+    }
+
+    fn initialization_bytecode(is_oog: bool) -> Bytecode {
+        let memory_bytes = [0x60; 10];
+        let memory_value = Word::from_big_endian(&memory_bytes);
+        let code_len = if is_oog { 0 } else { MAXCODESIZE + 1 };
```

Comment:
Suggest not to assign `0` when `is_oog` is true and you could remove `if-else` in L161, just assign `code_len`.

Reply:
if use code_len = MAXCODESIZE + 1  current test will met OOG(STATIC_EXPANSION) error which not implemented in this repo.

Reply:
Sorry, I didn't get your point. In your test, `tx_deploy_max_code_size_exceed`, you also sets `is_oog` to false. My suggestion is like 
```
let code_len = if is_oog { 5 } else { MAXCODESIZE + 1 };
```
then, you could replace 
```
PUSH2(if is_oog { 5 } else { code_len}) // length to copy
```
with
```
PUSH2(code_len)
```

Reply:
the point is when `is_oog` == false, the `code_len = MAXCODESIZE + 1`, which is very big.  `mstore` in following codes will  meet out of gas. 
           ```
 PUSH10(memory_value)
            PUSH32(code_len)
            MSTORE
```
geth error OutOfGas(StaticMemoryExpansion) occurred in  MSTORE at pc 44
thread 'evm_circuit::execution::error_code_store::test::tx_deploy_code_store_oog' panicked at 'evm_unimplemented: TODO: error state OutOfGas(StaticMemoryExpansion) not implemented', bus-mapping/src/evm/opcodes.rs:315:13
```

Reply:
I totally understood `code_len` is too big to `mstore`. 
However, in your current implementation, `is_oog` could be false as well, like your test `tx_deploy_max_code_size_exceed`. Which means in your current implementation, `code_len` could be `MAXCODESIZE + 1` as well. Wondering how do you pass in your current implemetation? 

Besides, my suggestion is for the case when `is_oog` is true, not for `is_oog` is false case.

Reply:
the method `initialization_bytecode` is used by oog and non oog case.  for non oog case `code_len` could be `MAXCODESIZE + 1`, the tx gas adjust to higher value  in method `test_context` as following. for oog case, the gas setting is low. so both of cases can be constructed.
```
|mut txs, accs| {
                txs[0]
                    .from(accs[0].address)
                    .to(accs[1].address)
                    .gas(if is_oog {
                        53800u64.into()
                    } else {
                        103800u64.into()
                    });
```

Reply:
This comment is just a minor suggestion in the beginning to make code more clean. 

I found the reason why it occurs `OutOfGas(StaticMemoryExpansion)` error after I modified to 
```
        let code_len = if is_oog { 5 } else { MAXCODESIZE + 1 };

        let mut code = bytecode! {
            PUSH10(memory_value)
            PUSH32(code_len)
            MSTORE
            PUSH2(code_len) // length to copy
            PUSH2(32u64 - u64::try_from(memory_bytes.len()).unwrap()) // offset
        };
```

If I updated to above snippet, only `tx_deploy_code_store_oog` fails. In this test case, it's not using `text_context` but using it's own created context with gas `53446u64`. If I updated the gas to `53800u64` (the same value as `test_context` given), above snippet would work well. It seems insufficient gas supply causes the `OutOfGas(StaticMemoryExpansion)` error.

Reply:
totally got, that's OK !

Reply:
updated in [0164da8](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1545/commits/0164da8db01be616f0764768e2652dcf03b7d28c) thanks!!